File synchronisation

benjay10 opened the pull request:
This should make file synchronisation work. All problems should be tackled (look at later commits).

Commit:
Added some tool files. Changed POST to GET for sync start.

Commit:
Ingested triples are now temporarily stored in a separate graph.

Commit:
Files are downloaded into the consumer, triples are then inserted in the ingest graph

Commit:
File synch should be nicely working. Some fine tuning and error handling later.

Commit:
Downloading files depends on the creation of a writeStream.

If writeStream can not be created, the file already exists and does not need to be downloaded again.

Commit:
Added JSDoc for most of the function dealing with file sync.

Commit:
Added storing of errors to the database

It stores, in addition to the error message, the stack trace of the
error.

Commit:
Files deleted after all metadata is deleted instead of as soon as possible.

Commit:
Attempts and simplified

Allow for multiple download attempts. After threshold, the metadata is
marked as failure and won't be tried again. Synchronisation is now done
in steps that can be executed in any order. They are more resilient to
errors and keep working.

Commit:
File remapping now working.

You can supply a remapping object in the config to map an original
folder structure onto a new one if required.

Commit:
Using the URI instead of filename

It now uses the URI of the physical file instead of the filename from the metadata to put the file in place directly instead of having to rename and move it after downloading.

Commit:
Added scope id to isolate updates to the temp graphs

Commit:
Completed the JSDoc

Commit:
Correct import problem, enabled automatic sync

cecemel commented on .vimrc:
```diff
@@ -0,0 +1,6 @@
+set tabstop=2
```
Best to move this file to .gitignore, and don't commit it.
Better to use .editorconfig, but keep this out of this PR, to keep the PR focused.

cecemel commented on Dockerfile:
```diff
@@ -1,2 +1,2 @@
-FROM semtech/mu-javascript-template:1.5.0-beta.2
-LABEL maintainer="Redpencil <info@redpencil.io>"
\ No newline at end of file
+FROM semtech/mu-javascript-template:1.5.0-beta.3
```
Even though it makes sense, keep this out of this PR, to keep the PR focused

benjay10 replied:
I want the .3 version locally, but revert to the original line and make sure this line stops appearing in the pull request. How do I do this on Git?

cecemel commented on FrequentSPARQLqueries.sparql:
```diff
@@ -0,0 +1,64 @@
+PREFIX bks:     <http://mu.semte.ch/services/github/madnificent/book-service/books/>
```
I don't think this file belongs here

cecemel commented on app.js:
```diff
@@ -3,30 +3,40 @@ import { CronJob } from 'cron';
 
 app.get('/', function(req, res) {
-  res.send(`Hello, you have reached ${SERVICE_NAME}! I'm doing just fine :)`);
+  res.status(200).json({ result: `Hello, you have reached ${SERVICE_NAME}! I'm doing just fine :)` });
```
Even though it makes sense, keep this out of this PR, to keep the PR focused

cecemel commented on app.js:
```diff
@@ -3,30 +3,40 @@ import { CronJob } from 'cron';
  */
 
-app.post('/initial-sync-jobs', async function( _, res ){
+app.get('/initial-sync-jobs', async function( _, res ){
```
Better not changing the API like that, even though the impact is minor. And perhaps not relevant to this PR

benjay10 replied:
I needed to, to be able to test. If you want, I need to figure out how to not include this in the pull request. (See my previous question about how to do this.)

cecemel replied:
You should be able to do post requests with all http clients  e.g. `wget --post-data='' http://host/path`, curl has equivalents too.

benjay10 replied:
But not with one click in a browser. I have to leave this for a while.

cecemel replied:
Don't commit it then i'd say

benjay10 replied:
Right! Weird brainfart of mine.

cecemel commented on app.js:
```diff
@@ -3,30 +3,40 @@ import { CronJob } from 'cron';
 new CronJob(CRON_PATTERN_DELTA_SYNC, async function() {
   const now = new Date().toISOString();
   console.info(`Delta sync triggered by cron job at ${now}`);
+  //console.log("Delta sync not executed, uncomment the code for that in the CronJob");
```
Please remove

cecemel commented on app.js:
```diff
@@ -37,9 +47,15 @@ app.delete('/initial-sync-jobs', async function( _, res ){
 });
 
-app.post('/delta-sync-jobs', async function( _, res ){
+app.get('/delta-sync-jobs', async function( req, res ){
```
Better not changing the API like that, even though the impact is minor. And perhaps not relevant to this PR

cecemel commented on app.js:
```diff
@@ -37,9 +47,15 @@ app.delete('/initial-sync-jobs', async function( _, res ){
   res.send({ msg: 'Started delta sync job' });
 });
 
+app.get("/file-sync", async function (req, res) {
```
If needed, perhaps best to keep it with a POST request. To be consistent, and the fact it is not idempotent

cecemel commented on build.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
```
Best to leave this file out

cecemel commented on config.js:
```diff
@@ -25,21 +25,32 @@ export const SYNC_DATASET_PATH = process.env.SYNC_DATASET_PATH || '/datasets';
 // GRAPHS
-export const INGEST_GRAPH = process.env.INGEST_GRAPH || `http://mu.semte.ch/graphs/public`;
-export const JOBS_GRAPH = process.env.JOBS_GRAPH || 'http://mu.semte.ch/graphs/system/jobs';
+export const INGEST_GRAPH            = process.env.INGEST_GRAPH            || "http://mu.semte.ch/graphs/public";
```
Please keep outlining with what has previously been used

benjay10 replied:
Sorry, outlining does not make sense if the entire paragraph is not done the same way.

cecemel replied:
what do you mean? I think in general the aim would be something like `cont foo[space]=[space]'bar'`

benjay10 replied:
The outlining is done specifically to align on special characters like '=' and '||' to make code more easily scannable and to spot mistakes more easily. I'm not breaking indentation rules.

cecemel replied:
I agree there are benefits, but like everything this is a tradeoff. And here, the past dictates the future. Such outline hasn't been used in the organization and can't be introduced like that. This has to fit in a broader discussion of code-conventions. Keep in mind, you will not be the only one working on the code base, other people work on it too. Now it is a matter of integrating in the team and going with the flow. Such style changes won't be appreciated if you all of a sudden introduce this on your own.

cecemel commented on config.js:
```diff
@@ -85,5 +96,5 @@ export const PREFIXES = `
   PREFIX adms: <http://www.w3.org/ns/adms#>
   PREFIX nfo: <http://www.semanticdesktop.org/ontologies/2007/03/22/nfo#>
-  PREFIX dbpedia: <http://dbpedia.org/resource/>
+  PREFIX dbpedia: <http://dbpedia.org/ontology/>
```
Good catch, just a heads up, make sure such changes don't have unexpected side-effects. Such changes could always trigger the need for migrations.

benjay10 replied:
Can we safely merge this? Is there a need to do migrations at this point?

cecemel replied:
no i checked it, it will be fine. Just always a thing to consider in the future.

benjay10 replied:
OK good. :smile:

cecemel commented on lib/delta-sync/delta-file.js:
```diff
@@ -80,12 +86,29 @@ export default class DeltaFile {
+        await deleteTriplesInAllGraphs(regularDeletes);
+
+        //Store triples about files in a temporary graph, will process later
+        //console.log("WHILE INGESTING, this is the partioning of inserts: ", JSON.stringify(partitionedInserts));
```
Remove redundant comments

cecemel commented on lib/delta-sync/delta-file.js:
```diff
@@ -80,12 +86,29 @@ export default class DeltaFile {
-        console.log(`Deleting data in all graphs`);
-        await deleteTriplesInAllGraphs(deletes);
+        //Filter on file triples
+        let partitionedInserts = partition(inserts, change => change.graph.value === UPLOADED_FILES_GRAPH);
```
Are you partitioning the data based on the graph it comes from? If so, this is brittle.. We would have gone for the URI's that start with `share://` right?

benjay10 replied:
Yes, on the graph. Graph URI can be configured in the config.js. We could consider supplying an array of URI's for better flexibility.
Filtering on `share://` only filters for the physical file metadata. Virtual file metadata is used for downloading.
We could rely on a specific prefix for all virtual file metadata, but that needs to be configured on the producer and consumer both. This means filtering on at least 2 prefixes for the URI's of triples.
This considering, filtering on graph URI sounds less brittle to me.

cecemel replied:
On the graph is a problem anyway, because a graph might contain more data than files alone. So there is a possibility some triples will never be ingested with this approach (since not necessarily attached to a file). Can we try it out with `share://`?
In the background job, you can then perform the check whether the data is complete to proceed with downloading?

benjay10 replied:
All right

cecemel commented on lib/delta-sync/delta-file.js:
```diff
@@ -80,12 +86,29 @@ export default class DeltaFile {
+        let partitionedDeletes = partition(deletes, change => change.graph.value === UPLOADED_FILES_GRAPH);
+        let regularInserts     = partitionedInserts.fails;
+        let regularDeletes     = partitionedDeletes.fails;
+        let fileInserts        = partitionedInserts.passes;
```
Keep the outline in line with what has previously been used.

benjay10 replied:
But this is what I have added to the file. Is the outlining annoying like this?

cecemel replied:
These boils down to religious wars i guess :-) but now I would assume most of the IDE's used by the team are not configured to deal with this kind of outline. Basically, right now, it is a pain to write and navigate through. I have no idea whether there any smart IDE's that can deal with e.g. a change in the variable `regularDeletes` to `regularDeletesFromSourceData`, which would entail a re-align of the full block.

So guess here, we'll have to face the fact that the past dictates the present and go with the flow.

benjay10 replied:
It is so easy to do in Vim (even without plugins) that I cannot believe it isn't done more often. I've seen it in C code, and it is also rather common in Lisp variants.

cecemel replied:
I refer to my previous comments. It  is a matter of integrating.

cecemel commented on lib/utils.js:
```diff
@@ -1,4 +1,4 @@
-import { uuid, sparqlEscapeUri, sparqlEscapeString, sparqlEscapeDateTime } from 'mu';
+import * as mu from 'mu';
```
It can have value, it is a debate, but goes against what is used so far in other repos too. So at least, I would keep it out of this PR. (also the related changes there, else it clutters it.)

cecemel commented on taskell.md:
```diff
@@ -0,0 +1,76 @@
+## To Do
```
Keep this file out of the PR. If it is not internal related to the task, you can track it under section TODOs of readme.md

Commit:
Added new way to throw and log errors

Created StackableError to combine errors on different levels or to
collect them in a list on the same level, and nicely print them.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+  }
+
+  //The options for the http request to the producer
+  let options = {
```
It would be better to just append the UUID to the base url. Also url parsing is better with new URL(), and don't forget port 80 etc is not a priori correct.

benjay10 replied:
Was already fixed in a newer commit.

benjay10 replied:
Port 80 is the default if no port is specified. At least this makes it explicitly obvious in the code.

cecemel replied:
But what happens if you sync over https? 
This information is just included in the SYNC_BASE_URL and I don't see why you start potentially making a wrong decomposition of it.

benjay10 replied:
So is it OK if it uses http or https based on what the URL says?

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+ */
+async function downloadAndSaveFile(vuuid, filepath) {
+
+  if (!filepath.startsWith(FILE_FOLDER)) {
```
I don't understand this constraint.

benjay10 replied:
The goal here is to protect against a weird case where the URI demands the file to be placed in a certain folder that is not a folder in the shared volume. This means that the file service will not be able to find the file for serving and the file will not persist when recreating the Docker container.

benjay10 replied:
Hmm actually, this is somewhat unnecessary, because while remapping, files are moved to their final destination. However, if there is no remapping entry, the files stay where they are, so best to place them correctly in the first place raising an error when something might get lost. Better safe than sorry?

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+  return new Promise((resolve, reject) => {
+
+    //First make sure the destination path exists
+    let targetdir = path.dirname(filepath);
```
you can simplify this with mkdirSync I think

benjay10 replied:
Once started with promises, why not use them everywhere for consistency?

cecemel replied:
It just clutters the code, whereas there is no significant benefit 

Btw: where possible also use async/await syntax

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+      resolve();
+    });
+  })
+  .then(() => new Promise((resolve, reject) => {
```
A think this is a bit elaborate. Perhaps take a simple approach like: https://github.com/lblod/download-url-service/blob/60f6950be80bf024cb27e888e732c271b3c526ab/app.js#L360

benjay10 replied:
It's elaborate but protects against leaving empty files around and therefore hindering the proper download of the file later. Only, it's mixed with all the rest of file downloading.

cecemel replied:
This is how we do it in other services: https://github.com/lblod/download-url-service/blob/60f6950be80bf024cb27e888e732c271b3c526ab/app.js#L256

benjay10 replied:
But that is what I did, just spread out in different functions and using native http libraries.

cecemel replied:
Can we just try to stick with the suggestion I made? Again, you won't be the only one working on the code base. These are really a lot of lines for the next maintainer to parse. And every line of code is potentially a bug.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+
+  //Rename and move the files, remap the uri
+  try {
+    await remapFiles();
```
I am not sure why this can not be done on save file

benjay10 replied:
This was my effort to keep the remapping as a separate process that can fail 'gracefully'. So if the remapping fails, the file is still safe on the consumer, but in the wrong place. Different approach?

cecemel replied:
Not sure I see the benefit here. If you write the file immediately to the correct location, there is no need to introduce a second step which could potentially fail. I might miss something though

benjay10 replied:
Remapping also involves rewriting the URIs in the database. This can only be done once all the metadata has arrived, otherwise you would have to somehow track which triples are already remapped and which aren't. When the file is being downloaded, not all metadata has necessarily arrived. I wanted to isolate changing the URI (which contains the file path) and moving the file to that path in one step in order to avoid an inconsistent state.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+ * @async
+ * @function runFileSync
+ */
+async function runFileSync() {
```
It seems the operations on files are processed per batch. Is there no way to group these operations per file?
It seems now the error handling becomes too complex.
So something along the lines:
```
const vUris = getFilesToIngest()
for(vUri ofvUris){
try
  await publishFile(vUri)
catch
   // error handling
}
```
and the same for removing the files

benjay10 replied:
Error handling is now improved with StackableError. This will (hopefully) give useful error messages about what files failed, but still process them in batch as much as possible for better performance(?) and simplicity.

cecemel replied:
I don't see how StackableError can make it more simple. If you consider the operations on one single file as block in a function, you should have all the info required to elaborate on your error messages -if needed- without breaking the full loop.

benjay10 replied:
The StackableError is not breaking the loop. There is an error message per file, if any files failed processing.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+                dct:created           ?pcreated   ;
+                dct:modified          ?pmodified  .
+        }
+        GRAPH <${INGEST_GRAPH}> {
```
The assumption here would be that the file to delete has been ingested previously. I am unsure that we really want to delete files which have not been published earlier. There might be good reasons here, so curious.

benjay10 replied:
I don't really understand what you mean here.
Files need to be fully published first (because metadata is only published when all triples have arrived) before they can be removed. If you remove a file first, the insert wins and the removal had no effect. It all depends on timing, but usually, there is enough time between insert and removal of the same file to not cause a problem.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+      }
+    `;
+    let results = await querySudo(queryString);
+    let paths   = results.results.bindings.map(b => b.new2puri.value.replace("share://", "/share/"));
```
Try to keep outline consistent with previously used in repo

cecemel replied:
I would also try to make a difference between operations which access the database, and manipulation of the uris. So i would move this step out

benjay10 replied:
I've separated it.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+                nfo:fileName          ?filename   ;
+                dct:format            ?format     ;
+                nfo:fileSize          ?filesize   ;
+                dbpedia:fileExtension ?extension  ;
```
Is it possible to just use 2 spaces indent?

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+    DELETE {
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ?vuri a nfo:FileDataObject              ;
+              a ext:DownloadSuccess             ;
```
same 2 spaces request.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+      }
+    }
+    `;
+  //This is query is shorter, and more readable, but takes too long to execute, but is otherwise identical in functionality.
```
I fail to see why this would be slower. Are you sure there wasn't any unbound variable?

benjay10 replied:
(I'm a beginner with SPARQL)
The query that is used already take a couple of _seconds_. The one that is in comments took _55 seconds_, which caused time-outs and failures. I don't know why.

cecemel commented on lib/delta-sync/file-sync.js:
```diff
@@ -0,0 +1,1051 @@
+      GRAPH <${TEMP_FILE_GRAPH}> {
+        ?vuri a nfo:FileDataObject              ;
+              a ext:DownloadSuccess             ;
+              a ext:RemapSuccess                ;
```
Viewing this, I would suggest to introduce an ext:FileIngestSyncTask, ext:FileRemoveSyncTask (which granted, is not the nicest class)
But this would serve an overarching entity you can query on, attach statuses, download attempts, errors etc.

Let's step back and see what the model of such a task could be.

- has created & modified status (dct:created, dct:modified)
- has one status (adms:status)
- has zero or more errors (http://redpencil.data.gift/vocabularies/tasks/error)
- dct:subject: the physical file to process
- ext:numberOfRetries 

what could be the statuses?
- waiting for all data to come in (we could use TASK_NOT_STARTED_STATUS)
- processing  (TASK_ONGOING_STATUS)
- success/failure (TASK_SUCCESS_STATUS, TASK_FAILED_STATUS)

Based on this,  A periodic job to 
- query all tasks not started, and see whether all triples for a complete file are there
- process them
- if ok -> success,
  -  if nok and failed in the download procedure. set to task not started yet if max attempts have not been exceeded. 
  - If nok and somewhere else, fail the task immediately

benjay10 replied:
You realize that fully integrating tasks will necessitate a rewrite of most of the file synchronisation and will over complicate the process for very little benefit (in my opinion)?

cecemel commented on lib/delta-sync/delta-file.js:
```diff
@@ -80,12 +86,29 @@ export default class DeltaFile {
+        //Store triples about files in a temporary graph, will process later
+        //console.log("WHILE INGESTING, this is the partioning of inserts: ", JSON.stringify(partitionedInserts));
+        //console.log("WHILE INGESTING, this is the partioning of deletes: ", JSON.stringify(partitionedDeletes));
+        await insertTriplesInGraph(fileInserts, TEMP_FILE_GRAPH);
```
See also comment near line : https://github.com/lblod/delta-consumer-single-graph-maintainer/pull/1/files#diff-e3614c29bf20e554c84cb4e818d5b6bcd8da1d3708cbb6232c0a3cb81be6a162R430

about the specif task to create . This will make life easier i think

cecemel replied:
Also, once these triples are injected in their separate container, we provide the scopeId so other services can opt out. See also: https://gist.github.com/cecemel/62643906d0d0da66aadb851137775c43#problem-1-timing-and-internal-publishing-of-files the side bits part

benjay10 replied:
ScopeId already used elsewhere in the file-sync. Will add it here too!

cecemel commented on lib/delta-sync/delta-file.js:
```diff
@@ -80,12 +86,29 @@ export default class DeltaFile {
       for (let {inserts, deletes} of changeSets) {
-        console.log(`Deleting data in all graphs`);
-        await deleteTriplesInAllGraphs(deletes);
+        //Filter on file triples
```
I would like to have a big if/else block here. With the an environment to determine whether we need file sync. if We don't need filesync, I would fall back to the previous code, we know it works, as a safety measure. 
You can keep the change made on `insertTriplesInGraph`

Commit:
Added updates to metadata: naive push to ingest

Commit:
Slightly changed download of files. Don't want rewrite because it is
rather robust like this.

Commit:
Small fix on the promise chaining for file downloads

Commit:
Added some file to .gitignore

Commit:
Removed unGittable files

Commit:
Forgot about a random build file

Commit:
Undo some changes to clean up a pull request. Mostly aesthetic changes

Commit:
Keep URL transformation apart from database querying

Commit:
Removed some comments

Commit:
Filtering of file metadata triples is now done on URI prefix instead of graph.

Commit:
Using scope id on inserting triples in temp graph

Commit:
Reverting back to old app.js and including needed changes again

Commit:
Reverted Dockerfile

Commit:
Added conditional whether to perform separate file sync or not

Commit:
Forgotten import and switching between http and https based on url
